rpc/dialer.go

```diff
@@ -240,6 +241,156 @@ func DialDirectGRPC(ctx context.Context, address string, logger utils.ZapCompati
+		return nil, err
+	}
+	return conn, nil
+}
+
+// SocksProxyFallbackDialContext will return nil if SocksProxyEnvVar is not set or if trying to connect to a local address,
+// which will allow dialers to use the default DialContext.
+// If SocksProxyEnvVar is set, it will prioritize a connection made without a proxy but will fall back to a SOCKS proxy connection.
+func SocksProxyFallbackDialContext(
+	ctx context.Context,
+	network, addr string,
+	logger utils.ZapCompatibleLogger,
+) func(ctx context.Context, network, addr string) (net.Conn, error) {
+	// Use SOCKS proxy from environment as gRPC proxy dialer. Do not use SOCKS proxy if trying to connect to a local address.
+	localAddr := strings.HasPrefix(addr, "[::]") || strings.HasPrefix(addr, "localhost") || strings.HasPrefix(addr, "unix")
+	proxyAddr := os.Getenv(SocksProxyEnvVar)
+	if localAddr || proxyAddr == "" {
+		// return nil in these cases so that the default dialer gets used instead.
+		return nil
+	}
+
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		// if ONLY_SOCKS_PROXY specified, no need for a parallel dial - only dial through
+		// the SOCKS proxy directly.
+		if os.Getenv(OnlySocksProxyEnvVar) != "" {
+			logger.Infow("Both SOCKS_PROXY and ONLY_SOCKS_PROXY specified, only SOCKS proxy will be used for outgoing connection")
+			conn, err := socksProxyDialContext(ctx, network, proxyAddr, addr)
+			if err == nil {
+				logger.Infow("connected with SOCKS proxy")
+			}
+			return conn, err
+		}
+
+		// the block below heavily references https://go.dev/src/net/dial.go#L585
+		type dialResult struct {
+			net.Conn
+			error
+			primary bool
+			done    bool
+		}
+		results := make(chan dialResult) // unbuffered
+
+		var primary, fallback dialResult
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		// otherwise, do a parallel dial with a slight delay for the fallback option.
+		returned := make(chan struct{})
+		defer close(returned)
+
+		dialer := func(ctx context.Context, dialFunc func(context.Context) (net.Conn, error), primary bool) {
+			defer wg.Done()
+			conn, err := dialFunc(ctx)
+			select {
+			case results <- dialResult{Conn: conn, error: err, primary: primary, done: true}:
+			case <-returned:
+				if conn != nil {
+					utils.UncheckedError(conn.Close())
+				}
+			}
+		}
+
+		logger.Infow("SOCKS_PROXY specified, SOCKS proxy will be used as a fallback for outgoing connection")
+		// start the main dial attempt.
+		primaryCtx, primaryCancel := context.WithCancel(ctx)
+		defer primaryCancel()
+		wg.Add(1)
+		primaryDial := func(ctx context.Context) (net.Conn, error) {
+			// create a zero-valued net.Dialer to use net.Dialer's default DialContext method
+			var zeroDialer net.Dialer
+			return zeroDialer.DialContext(ctx, network, addr)
+		}
+		go dialer(primaryCtx, primaryDial, true)
+
+		// wait a small amount before starting the fallback dial (to prioritize the primary connection method).
+		fallbackTimer := time.NewTimer(time.Second)
```

Comment:
> time.Second

If I'm reading Golang source code correctly, their fallback delay is customizable and defaults to 300ms. I understand not exposing this as a knob, but any reason to bump to 1s? Did you see SOCKS being "over" prioritized with 300ms?

Reply:
nope, let me test and drop to 300ms if no issues